File: services/ingestion-service/internal/domain/valueobject/access_key.go
package valueobject

import (
	"errors"
	"unicode"
)

var ErrInvalidAccessKey = errors.New("access key must be exactly 44 digits")

// AccessKey is the NF-e 44-digit access key.
type AccessKey string

func NewAccessKey(raw string) (AccessKey, error) {
	if len(raw) != 44 {
		return "", ErrInvalidAccessKey
	}
	for _, c := range raw {
		if !unicode.IsDigit(c) {
			return "", ErrInvalidAccessKey
		}
	}
	return AccessKey(raw), nil
}

func (a AccessKey) String() string { return string(a) }

File: services/tax-calculator/src/domain/value-objects/access-key.ts
import { DomainError } from '../errors/domain-error.js';
import { type Result, ok, err } from '../result.js';

export class InvalidAccessKeyError extends DomainError {
  constructor(value: string) {
    super(`Access key must be 44 digits, got: "${value}"`);
  }
}

export class AccessKey {
  private constructor(private readonly _value: string) {}

  static create(value: string): Result<AccessKey, InvalidAccessKeyError> {
    const digits = value.replace(/\D/g, '');
    if (digits.length !== 44) {
      return err(new InvalidAccessKeyError(value));
    }
    return ok(new AccessKey(digits));
  }

  get value(): string {
    return this._value;
  }

  toString(): string {
    return this._value;
  }
}
